functional/lenses.md

```diff
@@ -0,0 +1,365 @@
+More powerful lenses can be created which both set and get data in a
+structure.
+But the concept really becomes interesting when used as a building block for
+composition.
+That is where the concept appears more clearly in Rust.
+
+## Prisms: A Higher-Order form of "Optics"
+
+A simple function such as `unique_ids` above operates on a single lens.
+A *prism* is a function that operates on a *family* of lenses.
+It is one conceptual level higher, using lenses as a building block, and
+continuing the metaphor, is part of a family of "optics".
+It is the main one that is useful in understanding Rust APIs, so will be the
+focus here.
+
+The same way that traits allow "lens-like" design with static polymorphism and
+dynamic dispatch, prism-like designs appear in Rust APIs which split problems
+into multiple associated types to be composed.
+A good example of this is the traits in the parsing crate Serde.
+
+Trying to understand the way Serde works by only reading the API is a
+challenge, especially the first time.
+Consider the `Deserializer` trait, implemented by some type in any library
+which parses a new format:
+
+```rust,ignore
+pub trait Deserializer<'de>: Sized {
+    type Error: Error;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>;
+
+    fn deserialize_bool<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>;
+
+    // remainder ommitted
+}
+```
+
+For a trait that's just supposed to parse data from a format and return a
+value, this looks odd.
+Why are all the return types type erased?
+
+To understand, keep the lens concept in mind and look at the definition of
+the `Visitor` type that is passed in generically:
+
+```rust,ignore
+pub trait Visitor<'de>: Sized {
+    type Value;
+
+    fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+    where
+        E: Error;
+
+    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+    where
+        E: Error;
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: Error;
+
+    // remainder omitted
+}
+```
+
+The job of the `Visitor` type is to construct values in the Serde data model,
+which are represented by its associated `Value` type.
+
+These values represent parts of the Rust value being deserialized.
+If this fails, it returns an `Error` type -- an error type determined by the
+`Deserializer` when its methods were called.
+
+This highlights that `Deserializer` is similar to `CustomerId` from earlier,
+allowing any format parser which implements it to create `Value`s based on what
+it parsed.
+The `Value` trait is acting like a lens in functional languages.
+
+But unlike the `CustomerId` trait, the return types of `Visitor` methods are
+*generic*, and the concrete `Value` type is *determined by the Visitor itself*.
+
+Instead of acting as one lens, it effectively a family of
+lenses, one for each concrete type of `Visitor`.
```

Comment:
```suggestion
Instead of acting as one lens, it effectively acts as a family of
lenses, one for each concrete type of `Visitor`.
```
Adding the verb